sdk/core/core-rest-pipeline/src/policies/auxiliaryAuthenticationPolicy.ts

```diff
@@ -0,0 +1,116 @@
+import { AuthorizeRequestOptions } from "./bearerTokenAuthenticationPolicy";
+
+/**
+ * The programmatic identifier of the bearerTokenAuthenticationPolicy.
+ */
+export const auxiliaryAuthenticationPolicyName = "auxiliaryAuthenticationPolicy";
+const AUTHORIZATION_AUXILIARY_HEADER = "x-ms-authorization-auxiliary";
+
+/**
+ * Options to configure the auxiliaryAuthenticationPolicy
+ */
+export interface AuxiliaryAuthenticationPolicyNameOptions {
+  /**
+   * The TokenCredential list that can supply the auxiliary authentication token.
+   */
+  credentials?: TokenCredential[];
+  /**
+   * The scopes for which the auxiliary token applies.
+   */
+  scopes: string | string[];
+  /**
+   * A logger can be sent for debugging purposes.
+   */
+  logger?: AzureLogger;
+}
+
+/**
+ * Default authorize request handler
+ */
+async function defaultAuthorizeRequest(options: AuthorizeRequestOptions): Promise<string | null> {
+  const { scopes, getAccessToken, request } = options;
+  const getTokenOptions: GetTokenOptions = {
+    abortSignal: request.abortSignal,
+    tracingOptions: request.tracingOptions,
+  };
+
+  const accessToken = await getAccessToken(scopes, getTokenOptions);
+
+  if (accessToken) {
+    return accessToken.token;
+  }
+  return null;
+}
+
+function buildAccessTokenCycler(credential: TokenCredential) {
+  return credential ? createTokenCycler(credential /* , options */) : () => Promise.resolve(null);
+}
+
+/**
+ * A policy that can request a token from a TokenCredential implementation and
+ * then apply it to the Authorization header of a request as a Bearer token.
+ */
+export function auxiliaryAuthenticationPolicy(
+  options: AuxiliaryAuthenticationPolicyNameOptions
+): PipelinePolicy {
+  const { credentials, scopes } = options;
+  const logger = options.logger || coreLogger;
+  const tokenCyclerMap = new Map();
+
+  return {
+    name: auxiliaryAuthenticationPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      if (!request.url.toLowerCase().startsWith("https://")) {
+        throw new Error(
+          "Auxiliary token authentication is not permitted for non-TLS protected (non-https) URLs."
+        );
+      }
+
+      const tokenList: string[] = [];
+      for (const credential of credentials ?? []) {
+        if (!tokenCyclerMap.has(credential)) {
+          tokenCyclerMap.set(credential, buildAccessTokenCycler(credential));
+        }
+        const getAccessToken = tokenCyclerMap.get(credential);
+        const singalAccessToken = await defaultAuthorizeRequest({
+          scopes: Array.isArray(scopes) ? scopes : [scopes],
+          request,
+          getAccessToken,
+          logger,
+        });
+        if (singalAccessToken) {
+          tokenList.push(singalAccessToken);
+        }
+      }
+
+      request.headers.set(
+        AUTHORIZATION_AUXILIARY_HEADER,
+        tokenList.map((token) => `Bearer ${token}`).join(",")
```

Comment:
again, what if we have more than three tokens? This will currently append all of them.

Reply:
I commented here, https://github.com/Azure/azure-sdk-for-js/pull/25270#discussion_r1176209436.

@xiangyan99 How do you think?

Reply:
If there are more than three tokens, it means there are some configuration issues. As a client library, I don't think we want to decide which tokens to ignore on behalf of users and try to swallow the issue. Instead, I prefer to let the request fail and ask users to check the configuration. @xirzec your ideas?